R/derive_param_computed.R

```diff
@@ -147,48 +196,243 @@
 #'   constant_parameters = c("HEIGHT"),
 #'   constant_by_vars = exprs(USUBJID)
 #' )
-derive_param_computed <- function(dataset,
+#'
+#' # Example 3: Using data from an additional dataset and other variables than AVAL
+#' qs <- tibble::tribble(
+#'   ~USUBJID, ~AVISIT,   ~QSTESTCD, ~QSORRES, ~QSSTRESN,
+#'   "1",      "WEEK 2",  "CHSF112", NA,               1,
+#'   "1",      "WEEK 2",  "CHSF113", "Yes",           NA,
+#'   "1",      "WEEK 2",  "CHSF114", NA,               1,
+#'   "1",      "WEEK 4",  "CHSF112", NA,               2,
+#'   "1",      "WEEK 4",  "CHSF113", "No",            NA,
+#'   "1",      "WEEK 4",  "CHSF114", NA,               1
+#' )
+#'
+#' adchsf <- tibble::tribble(
+#'   ~USUBJID, ~AVISIT,  ~PARAMCD, ~QSORRES, ~QSSTRESN, ~AVAL,
+#'   "1",      "WEEK 2", "CHSF12", NA,       1,             6,
+#'   "1",      "WEEK 2", "CHSF14", NA,       1,             6,
+#'   "1",      "WEEK 4", "CHSF12", NA,       2,            12,
+#'   "1",      "WEEK 4", "CHSF14", NA,       1,             6
+#' )
+#'
+#' derive_param_computed(
+#'   adchsf,
+#'   dataset_add = qs,
+#'   by_vars = exprs(USUBJID, AVISIT),
+#'   parameters = exprs(CHSF12, CHSF13 = QSTESTCD %in% c("CHSF113", "CHSF213"), CHSF14),
+#'   analysis_value = case_when(
+#'     QSORRES.CHSF13 == "Not applicable" ~ 0,
+#'     QSORRES.CHSF13 == "Yes" ~ 38,
+#'     QSORRES.CHSF13 == "No" ~ if_else(
+#'       QSSTRESN.CHSF12 > QSSTRESN.CHSF14,
+#'       25,
+#'       0
+#'     )
+#'   ),
+#'   set_values_to = exprs(PARAMCD = "CHSF13")
+#' )
+derive_param_computed <- function(dataset = NULL,
+                                  dataset_add = NULL,
                                   by_vars,
                                   parameters,
+                                  analysis_var = AVAL,
                                   analysis_value,
                                   set_values_to,
                                   filter = NULL,
                                   constant_by_vars = NULL,
                                   constant_parameters = NULL) {
   assert_vars(by_vars)
+  analysis_var <- assert_symbol(enexpr(analysis_var))
   assert_vars(constant_by_vars, optional = TRUE)
-  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, PARAMCD, AVAL))
+  assert_data_frame(dataset, required_vars = by_vars, optional = TRUE)
+  assert_data_frame(dataset_add, optional = TRUE)
   filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
-  params_available <- unique(dataset$PARAMCD)
-  assert_character_vector(parameters, values = params_available)
-  assert_character_vector(constant_parameters, values = params_available, optional = TRUE)
   assert_varval_list(set_values_to)
   if (!is.null(set_values_to$PARAMCD)) {
     assert_param_does_not_exist(dataset, set_values_to$PARAMCD)
   }
+  analysis_value <- enexpr(analysis_value)
+
+  parameters <- assert_parameters_argument(parameters)
```

Comment:
should this be added to `get_hori_data` as well?
`get_hori_data` is not working when `parameters` is a character vector -

```r
advs <- tribble(
  ~USUBJID,      ~PARAMCD, ~PARAM,                            ~AVAL, ~AVALU, ~VISIT,
  "01-701-1015", "DIABP",  "Diastolic Blood Pressure (mmHg)",    51, "mmHg", "BASELINE",
  "01-701-1015", "DIABP",  "Diastolic Blood Pressure (mmHg)",    50, "mmHg", "WEEK 2",
  "01-701-1015", "SYSBP",  "Systolic Blood Pressure (mmHg)",    121, "mmHg", "BASELINE",
  "01-701-1015", "SYSBP",  "Systolic Blood Pressure (mmHg)",    121, "mmHg", "WEEK 2",
  "01-701-1028", "DIABP",  "Diastolic Blood Pressure (mmHg)",    79, "mmHg", "BASELINE",
  "01-701-1028", "DIABP",  "Diastolic Blood Pressure (mmHg)",    80, "mmHg", "WEEK 2",
  "01-701-1028", "SYSBP",  "Systolic Blood Pressure (mmHg)",    130, "mmHg", "BASELINE",
  "01-701-1028", "SYSBP",  "Systolic Blood Pressure (mmHg)",    132, "mmHg", "WEEK 2"
)

get_hori_data(
  advs,
  by_vars = exprs(USUBJID, VISIT),
  parameters = c("SYSBP", "DIABP"),
  analysis_value = expr((AVAL.SYSBP + 2 * AVAL.DIABP) / 3),
  filter = NULL
)
```
```r
$hori_data
NULL

Warning message:
The input dataset does not contain any observations fullfiling the filter condition (NULL) for the parameter codes (PARAMCD) `"SYSBP"` and `"DIABP"`
No new observations were added. 
```

Reply:
Yes, I have added assertions to `get_hori_data()`.

Reply:
should  `assert_parameters_argument(parameters)` be changed to `parameters <- assert_parameters_argument(parameters)`?

Reply:
Yes, good point! I missed that.